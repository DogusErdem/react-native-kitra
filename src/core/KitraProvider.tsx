import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
