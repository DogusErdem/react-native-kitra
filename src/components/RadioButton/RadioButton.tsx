import { useState } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';
import Animated, { FadeIn, FadeOut } from 'react-native-reanimated';
import type { FCCWD, RadioButtonProps } from '../../types';
import { applyDefaults } from '../../core/KitraProvider';

const RadioButton: FCCWD<RadioButtonProps> = ({ theme, onChange, testID, style }) => {
  const [value, setValue] = useState(false);
  return (
    <TouchableOpacity
      activeOpacity={0.5}
      onPress={() => {
        setValue(prev => !prev);
        onChange(!value);
      }}
      testID={testID}
      style={[value ?
        { borderColor: theme?.primary, backgroundColor: theme?.white }
        :
        { borderColor: theme?.disabledDark, backgroundColor: theme?.disabledLight }, styles.container, style]}
    >
      {value && <Animated.View entering={FadeIn} exiting={FadeOut} style={[{ backgroundColor: theme?.primary }, styles.innerCircle]} />}
    </TouchableOpacity>
  );
};
export default applyDefaults(RadioButton);

const styles = StyleSheet.create({
  container: { height: 20, width: 20, borderWidth: 1, alignItems: 'center', justifyContent: 'center', borderRadius: 10 },
  innerCircle: { height: 12, width: 12, borderRadius: 10 },
});
