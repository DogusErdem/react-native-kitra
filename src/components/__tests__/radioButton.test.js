import { fireEvent, render, act } from '@testing-library/react-native';
import { Text } from 'react-native';
import { KitraProvider } from '../../core/KitraProvider';
import RadioButton from '../RadioButton/RadioButton';

/* eslint-disable no-undef */
const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);

describe('RadioButton', () => {
  it('default render', () => {
    const tree = renderWithContext(<RadioButton />);
    expect(tree).toMatchSnapshot();
  });
  it('render with onChange prop', () => {
    const onChange = jest.fn();
    const { getByTestId } = renderWithContext(<RadioButton testID="radioButton" onChange={onChange} />);
    const radioButton = getByTestId('radioButton');
    fireEvent.press(radioButton);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(true);
    fireEvent.press(radioButton);
    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenCalledWith(false);
  });
  it('render with custom style', () => {
    const { getByTestId } = renderWithContext(<RadioButton
      testID="radioButton"
      style={{ marginTop: 10, padding: 5, margin: 10 }}
    />);
    const radioButton = getByTestId('radioButton');
    expect(radioButton).toHaveStyle({ marginTop: 10, padding: 5, margin: 10 });
  });
  describe('render with custom theme', () => {
    it('render with default custom theme', () => {
      const { getByTestId } = renderWithContext(<RadioButton
        testID="radioButton"
        theme={{ default: { background: 'white', border: 'blue', dot: 'blue' } }}
      />);
      const radioButton = getByTestId('radioButton');
      expect(radioButton).toHaveStyle({ backgroundColor: 'white', borderColor: 'blue' });
    });
    it('render with active custom theme', () => {
      const { getByTestId } = renderWithContext(<RadioButton
        testID="radioButton"
        theme={{ active: { background: 'blue', border: 'white', dot: 'white' } }}
      />);
      const radioButton = getByTestId('radioButton');
      fireEvent.press(radioButton);
      const dot = getByTestId('radioDot');
      expect(radioButton).toHaveStyle({ backgroundColor: 'blue', borderColor: 'white' });
      expect(dot).toHaveStyle({ backgroundColor: 'white' });
    });
  });
});
