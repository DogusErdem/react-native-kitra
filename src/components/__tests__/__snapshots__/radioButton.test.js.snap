// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton default render 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#EFF0F3",
        "borderColor": "#C9D5EF",
        "borderRadius": 10,
        "borderWidth": 1,
        "height": 20,
        "justifyContent": "center",
        "opacity": 1,
        "width": 20,
      }
    }
  />
</View>
`;
